### PR TITLE
[opt-remark] Add support for emitting remarks about runtime casts.

### PR DESCRIPTION
**Pull request by gottesmm**
At a high level this required me to do a few things:

1. I added support to opt-remarks for inferring Decls from debug_value_addr uses as well. To do this I created an abstract composition class on top of all DebugInsts to make me able to write code with either of them.
2. I refactored isSingleInitAllocStack into getSingleInitAllocStackUse. The algorithm is the same, we just return the single use that we found instead of a bool.
3. I added support to getSingleInitAllocStackUse to handle the relevant casts if our value is the source. In such a case, this is just a destroy_addr of the value. I checked all uses of the old API and found we had one and that one did not look at whether a destroy was a destroy_addr or not... it just used it to determine lifetime with respect to optimizing a different value.

With that in hand, it was easy to just wire up checked_cast_addr_br/unconditional_checked_cast_addr in the opt remark visitor and use those APIs.

### CODE REVIEW

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
preset=buildbot_incremental,tools=RA,stdlib=RD,smoketest=macosx
@swift-ci Please test with preset macOS

**Comment by gottesmm** [comment]
Since this is only tested with a no-asserts stdlib, kicking off a run using the custom preset job!

**Comment by gottesmm** [comment]
@swift-ci test windows platform

**Comment by gottesmm** [comment]
@swift-ci test macOS platform

File: include/swift/SIL/DebugUtils.h [hunk 1]
```diff
@@ -218,6 +218,70 @@ inline SILBasicBlock::iterator eraseFromParentWithDebugInsts(
 bool hasNonTrivialNonDebugTransitiveUsers(
     PointerUnion<SILInstruction *, SILArgument *> V);
 
+/// A high level composition struct on top of debug value instructions that
+/// enable one to write abstract code over all debug insts.
+struct DebugInst {
+  enum class Kind {
+    Invalid = 0,
+    ObjectValue,
+    AddressValue,
```
Comment (adrian-prantl) [inline include/swift/SIL/DebugUtils.h:227]: I think Value and Address would be more consistent with the usual terminology.

Comment (gottesmm) [reply include/swift/SIL/DebugUtils.h:227]: I made it more specific in a new version I am finishing up real quick.

**Comment by gottesmm** [comment]
@swift-ci test

**Comment by gottesmm** [comment]
preset=buildbot_incremental,tools=RA,stdlib=RD,smoketest=macosx
@swift-ci Please test with preset macOS

**Comment by gottesmm** [comment]
@swift-ci test Linux platform

**Comment by gottesmm** [comment]
@swift-ci test Windows platform

**Review by atrick — APPROVED** [review]
lgtm, although I didn't have time to think about `DebugVarCarryingInst`

The singleInitAllocStack stuff is all just hacks to get extra verification to work anyway.

**Comment by gottesmm** [comment]
Linux hit another unrelated error.

**Comment by gottesmm** [comment]
@swift-ci test linux platform

**Pull request merged**